samza-api/src/main/java/org/apache/samza/task/TaskContext.java

```diff
@@ -54,4 +54,14 @@
    *
    */
   void setStartingOffset(SystemStreamPartition ssp, String offset);
+
+  /**
+   * Method to allow user to return customized context
+   *
+   * @param <T>  the type of user-defined task context
+   * @return  user-defined task context object
+   */
+  default <T> T getUserDefinedContext() {
```

Comment:
We have no way to specify the type T. This should be Object. FWIW, I tried this with IntelliJ and the compiler complained that there was no way to convert the returned value to type T. We only get away with `null` here because the null type is castable to any reference type. Do we have a test for this?